consensus/hotstuff/timeoutaggregator/timeout_collectors.go

```diff
@@ -0,0 +1,134 @@
+package timeoutaggregator
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/consensus/hotstuff"
+	"github.com/onflow/flow-go/module/mempool"
+)
+
+// NewCollectorFactoryMethod is a factory method to generate a TimeoutCollector for concrete view
+type NewCollectorFactoryMethod = func(view uint64) (hotstuff.TimeoutCollector, error)
+
+// TimeoutCollectors implements management of multiple timeout collectors indexed by view.
+// Implements hotstuff.TimeoutCollectors interface. Creating a TimeoutCollector for a
+// particular view is lazy (instances are created on demand).
+// This structure is concurrently safe.
+// TODO: once VoteCollectors gets updated to stop managing worker pool we can merge VoteCollectors and TimeoutCollectors using generics
+type TimeoutCollectors struct {
+	log                zerolog.Logger
+	lock               sync.RWMutex
+	lowestRetainedView uint64                               // lowest view, for which we still retain a TimeoutCollector and process timeouts
+	collectors         map[uint64]hotstuff.TimeoutCollector // view -> TimeoutCollector
+	createCollector    NewCollectorFactoryMethod            // factory method for creating collectors
+}
+
+var _ hotstuff.TimeoutCollectors = (*TimeoutCollectors)(nil)
+
+func NewTimeoutCollectors(log zerolog.Logger, lowestRetainedView uint64, createCollector NewCollectorFactoryMethod) *TimeoutCollectors {
+	return &TimeoutCollectors{
+		log:                log,
+		lowestRetainedView: lowestRetainedView,
+		collectors:         make(map[uint64]hotstuff.TimeoutCollector),
+		createCollector:    createCollector,
+	}
+}
+
+// GetOrCreateCollector retrieves the hotstuff.TimeoutCollector for the specified
+// view or creates one if none exists.
+//  -  (collector, true, nil) if no collector can be found by the view, and a new collector was created.
+//  -  (collector, false, nil) if the collector can be found by the view
+//  -  (nil, false, error) if running into any exception creating the timeout collector state machine
+// Expected error returns during normal operations:
+//  * mempool.DecreasingPruningHeightError - in case view is lower than lowestRetainedView
```

Comment:
I feel it should be legal for `TimeoutCollectors` to return a  `model.ErrViewForUnknownEpoch`:

```suggestion
	//  * mempool.DecreasingPruningHeightError if view is below the pruning threshold
	//  * model.ErrViewForUnknownEpoch if view is not yet pruned but no epoch containing the given view is known
```